executor/cook/executor.py

```diff
@@ -0,0 +1,363 @@
+        Grace period before forceful kill
+
+    Returns
+    -------
+    True if the process was killed, False if it terminated naturally.
+    """
+    process, _, _ = process_info
+
+    while is_running(process):
+
+        if stop_signal.isSet():
+            logging.info('Executor has been instructed to terminate')
+            kill_task(process_info, shutdown_grace_period_ms)
+            break
+
+        time.sleep(cook.RUNNING_POLL_INTERVAL_SECS)
+
+
+def get_task_state(exit_code):
+    """Interprets the exit_code and return the corresponding task status string
+
+    Parameters
+    ----------
+    exit_code: int
+        An integer that represents the return code of the task.
+
+    Returns
+    -------
+    A task status string corresponding to the exit code.
+    """
+    if exit_code > 0:
+        return cook.TASK_FAILED
+    elif exit_code < 0:
+        return cook.TASK_KILLED
+    else:
+        return cook.TASK_FINISHED
+
+
+def manage_task(driver, task, stop_signal, completed_signal, config, stdout_name, stderr_name):
+    """Manages the execution of a task waiting for it to terminate normally or be killed.
+       It also sends the task status updates, sandbox location and exit code back to the scheduler.
+       Progress updates are tracked on a separate thread and are also sent to the scheduler.
+       Setting the stop_signal will trigger termination of the task and associated cleanup.
+
+    Returns
+    -------
+    Nothing
+    """
+    task_id = get_task_id(task)
+    try:
+        # not yet started to run the task
+        update_status(driver, task_id, cook.TASK_STARTING)
+
+        sandbox_message = json.dumps({'sandbox-location': config.sandbox_location, 'task-id': task_id})
+        send_message(driver, sandbox_message, config.max_message_length)
+
+        process_info = launch_task(task, stdout_name, stderr_name)
+        if process_info:
+            # task has begun running successfully
+            update_status(driver, task_id, cook.TASK_RUNNING)
+        else:
+            # task launch failed, report an error
+            logging.info('Error in launching task')
+            update_status(driver, task_id, cook.TASK_ERROR)
+            return
+
+        process, _, _ = process_info
+        progress_watcher = cp.ProgressWatcher(config, completed_signal)
+        progress_updater = cp.ProgressUpdater(driver, task_id, config.max_message_length,
+                                              config.progress_sample_interval_ms, send_message)
+        progress_complete_event = cp.launch_progress_tracker(progress_watcher, progress_updater)
+
+        await_process_completion(stop_signal, process_info, config.shutdown_grace_period_ms)
+        exit_code = process.returncode
+
+        # set the completed signal to trigger progress updater termination and await progress
+        # updater termination if executor is terminating normally
+        completed_signal.set()
+        if not stop_signal.isSet():
```

Comment:
Should we not wait for `progress_complete_event` unconditionally, given that the final progress/status might be important?